[WIP][DataLoader] CollateIterableDataset prototype

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49363 [WIP][DataLoader] Prototype of SamplerIterableDataset
* #49186 [WIP][DataLoader] Prototype of BatchIterableDataset
* **#48933 [WIP][DataLoader] CollateIterableDataset prototype**

Prototype for CollateIterableDataset.
Move `collate_batch_fn` to BatchIterableDataset

- CollateIterableDataset
  - [x] Prototype
  - [x] Tests
- BatchIterableDataset
  - [x] Prototype
  - [x] Tests
- SamplerIterableDataset
  - [x] Prototype
  - [x] Tests

Differential Revision: [D25623635](https://our.internmc.facebook.com/intern/diff/D25623635)

ejguan committed:
[WIP] CollateDataset prototype

[ghstack-poisoned]

ejguan committed:
Update on "[WIP] CollateDataset prototype"


Prototype for CollateDataset.
~Problems:~
How to let DataLoader knows if there is CollateDataset in the pipeline?
One possible solution:
```python
setattr(IterableDataset, "__collate__", True)
```
All the derived class of IterableDataset after CollateDataset created will have this attribute, but it's not safe and reasonable.

Another solution, DataLoader factory the pipeline of Datasets with explicit call for CollateDataset.

Note: User should know CollateDataset exists in the pipeline.

[ghstack-poisoned]

ejguan committed:
Update on "[WIP] CollateDataset prototype"


Prototype for CollateDataset.
~Problems:~
How to let DataLoader knows if there is CollateDataset in the pipeline?
One possible solution:
```python
setattr(IterableDataset, "__collate__", True)
```
All the derived class of IterableDataset after CollateDataset created will have this attribute, but it's not safe and reasonable.

Another solution, DataLoader factory the pipeline of Datasets with explicit call for CollateDataset.

Note: User should know CollateDataset exists in the pipeline.

[ghstack-poisoned]

ejguan committed:
Update on "[WIP] CollateDataset prototype"


Prototype for CollateDataset.
Move `collate_batch_fn` to BatchDataset

- CollateDataset
  - [x] Prototype
  - [ ] Tests
- BatchDataset
  - [ ] Prototype
  - [ ] Tests
- PaddedBatchDataset
  - [ ] Prototype
  - [ ] Tests

Other:
- [ ] UnBatchDataset
- [ ] BucketBatchDataset
- [ ] SampleDataset

[ghstack-poisoned]

ejguan committed:
Update on "[WIP] CollateDataset prototype"


Prototype for CollateDataset.
Move `collate_batch_fn` to BatchDataset

- CollateDataset
  - [x] Prototype
  - [x] Tests
- BatchDataset
  - [x] Prototype
  - [x] Tests
- SampleDataset
  - [ ] Prototype
  - [ ] Tests

Other:
- [ ] UnBatchDataset
- [ ] BucketBatchDataset
- [ ] PaddedBatchDataset

[ghstack-poisoned]

ejguan commented on torch/utils/data/datasets/collatedataset.py:
```diff
@@ -0,0 +1,65 @@
+        for data in self.dataset:
+            yield self.collate_fn(data)
+
+    # `__len__` is attached to class not instance
+    # Assume dataset has implemented `__len__` or raise NotImplementedError
+    def __len__(self) -> int:
+        if isinstance(self.dataset, Sized) and len(self.dataset) >= 0:
+            return len(self.dataset)
+        else:
+            raise NotImplementedError
```
As it's possible `dataset` did not implement `__len__`, we have to check here.
However, it creates another interesting problem has been discussed here:
https://github.com/pytorch/pytorch/blob/6820745e28e6c777d35a2b874b3e5b627e58c3a0/torch/utils/data/sampler.py#L39-L44
`list(dataset)` will try to call `__len__` first.If it doesn't work, it will try to use `__iter__` to create the list. 
Our current implementation prevents this behavior.

ejguan replied:
Besides, we may need to consider both `Infinite` size and invalid size (not implemented).

glaringlee replied:
@ejguan agree. Currently, the `length` in container dataset (listdir, tar, etc.) is a nominal length which is adopted from Nvidia webdataset, need to carefully define it.

ejguan replied:
We may need to add another layer above `IterableDataset` to define all these behavior.

ejguan committed:
Update on "[WIP][DataLoader] CollateIterableDataset prototype"


Prototype for CollateIterableDataset.
Move `collate_batch_fn` to BatchIterableDataset

- CollateIterableDataset
  - [x] Prototype
  - [x] Tests
- BatchIterableDataset
  - [x] Prototype
  - [x] Tests
- SamplerIterableDataset
  - [x] Prototype
  - [x] Tests

Other:
- [ ] UnBatchDataset
- [ ] BucketBatchDataset
- [ ] PaddedBatchDataset

[ghstack-poisoned]

ejguan committed:
Update on "[WIP][DataLoader] CollateIterableDataset prototype"


Prototype for CollateIterableDataset.
Move `collate_batch_fn` to BatchIterableDataset

- CollateIterableDataset
  - [x] Prototype
  - [x] Tests
- BatchIterableDataset
  - [x] Prototype
  - [x] Tests
- SamplerIterableDataset
  - [x] Prototype
  - [x] Tests

Other:
- [ ] UnBatchDataset
- [ ] BucketBatchDataset
- [ ] PaddedBatchDataset

[ghstack-poisoned]

ejguan committed:
Update on "[WIP][DataLoader] CollateIterableDataset prototype"


Prototype for CollateIterableDataset.
Move `collate_batch_fn` to BatchIterableDataset

- CollateIterableDataset
  - [x] Prototype
  - [x] Tests
- BatchIterableDataset
  - [x] Prototype
  - [x] Tests
- SamplerIterableDataset
  - [x] Prototype
  - [x] Tests

Other:
- [ ] UnBatchDataset
- [ ] BucketBatchDataset
- [ ] PaddedBatchDataset

[ghstack-poisoned]

ejguan committed:
Update on "[WIP][DataLoader] CollateIterableDataset prototype"


Prototype for CollateIterableDataset.
Move `collate_batch_fn` to BatchIterableDataset

- CollateIterableDataset
  - [x] Prototype
  - [x] Tests
- BatchIterableDataset
  - [x] Prototype
  - [x] Tests
- SamplerIterableDataset
  - [x] Prototype
  - [x] Tests

Differential Revision: [D25623635](https://our.internmc.facebook.com/intern/diff/D25623635)

[ghstack-poisoned]

ejguan committed:
Update on "[WIP][DataLoader] CollateIterableDataset prototype"


Prototype for CollateIterableDataset.
Move `collate_batch_fn` to BatchIterableDataset

- CollateIterableDataset
  - [x] Prototype
  - [x] Tests
- BatchIterableDataset
  - [x] Prototype
  - [x] Tests
- SamplerIterableDataset
  - [x] Prototype
  - [x] Tests

Differential Revision: [D25623635](https://our.internmc.facebook.com/intern/diff/D25623635)

[ghstack-poisoned]
